### ISSUE TITLE
torch.randn on CPU is not accurate with dtype=bfloat16

### ISSUE THREAD

**Issue by constroy** [pytorch/pytorch#166853]
### 🐛 Describe the bug

`torch.randn(dtype=torch.bfloat16)` on CPU only generates values in range (-3.5, +3.5). 
While that on CUDA can generate values in a wider range.

Run this python code below.
No values smaller than -3.5 will be found on CPU, but some will be found on CUDA.

```python
import torch                                                                                                                                                                                                                                                                  

dtype=torch.bfloat16

for i in range(10000):
    a = torch.randn((64, 64), dtype=dtype, device='cpu')
    if (a < -3.5).any(): # never True
        print('found a < -3.5 on CPU', a.min())
        break


for i in range(10000):
    a = torch.randn((64, 64), dtype=dtype, device='cuda')
    if (a < -3.5).any():
        print('found a < -3.5 on CUDA', a.min())
        break

```

I guess the cpu kernel should use a higher precision (float or double) for calculation.

https://github.com/pytorch/pytorch/blob/defac66e39c60f867795346ded171366ed682ee3/aten/src/ATen/native/cpu/DistributionTemplates.h#L139-L148

### Versions

Collecting environment information...
PyTorch version: 2.7.1
Is debug build: False
CUDA used to build PyTorch: 12.9
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.29.1
Libc version: glibc-2.31

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-2.4.3.3.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB

Nvidia driver version: 535.129.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.10.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.10.2
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          256
On-line CPU(s) list:             0-255
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           1
Model name:                      AMD EPYC 7763 64-Core Processor
Stepping:                        1
Frequency boost:                 enabled
CPU MHz:                         3094.157
CPU max MHz:                     2450.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        4890.76
Virtualization:                  AMD-V
L1d cache:                       4 MiB
L1i cache:                       4 MiB
L2 cache:                        64 MiB
L3 cache:                        512 MiB
NUMA node0 CPU(s):               0-63,128-191
NUMA node1 CPU(s):               64-127,192-255
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate sme ssbd mba sev ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cudnn-frontend==1.12.0
[pip3] nvtx==0.2.12
[pip3] pytorch-lightning==2.5.2
[pip3] torch==2.7.1
[pip3] torch-tb-profiler==0.4.3
[pip3] torchcde==0.2.5
[pip3] torchcfm==1.0.7
[pip3] torchdiffeq==0.2.5
[pip3] torchdyn==1.0.6
[pip3] torchmetrics==1.7.4
[pip3] torchsde==0.2.6
[pip3] torchvision==0.21.0+7af6987
[pip3] triton==3.3.1
[conda] Could not collect


cc @pbelevich

**Comment by constroy** [comment]
> And what GPU does in this case?

I tested this on Nvidia A100.
I think the code behaves as expected on GPU, but not on **CPU**.

**Comment by PPG11** [comment]
Thanks @constroy for tracking this down — we hit exactly this. Noise for
diffusion training was sampled on CPU in `bfloat16` while the sampling loop runs
on CUDA; because the CPU kernel does Box-Muller in the output dtype, the
training-side noise had its tails clipped at ~+-3.5 while the CUDA side did not.
Mean and std still look right, so nothing flags it — the "correctness (silent)"
label is very fitting.

Workaround, in case it helps others: sample in fp32 and cast.

```python
noise = torch.randn(shape, dtype=torch.float32, generator=g).to(torch.bfloat16)
```

This is a drop-in swap: per `normal_fill`, the uniform pre-fill consumes one
`random()` per element for both dtypes, so the generator state is not shifted and
seeded reproducibility elsewhere is preserved.

+1 for doing the transform in the accumulate type like the CUDA kernel already
does, so CPU and CUDA agree for the same dtype.

**Comment by constroy** [comment]
There is a simple fix in #166854.
Could any one review or proceed it? @malfet
